cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -684,6 +694,60 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             ctx.emit(Inst::setcc(cc, dst));
         }
 
+        Opcode::Fcmp => {
+            let condcode = inst_fp_condcode(ctx.data(insn)).unwrap();
+
+            let lhs = input_to_reg_mem(ctx, inputs[0]);
+            let rhs = input_to_reg(ctx, inputs[1]);
+            let dst = output_to_reg(ctx, outputs[0]);
+
+            // Unordered is returned by setting ZF, PF, CF <- 111
+            // Greater than by ZF, PF, CF <- 000
+            // Less than by ZF, PF, CF <- 001
+            // Equal by ZF, PF, CF <- 100
+
+            // Because setcc check together the CF and ZF, we swap
+            // operands for condition codes that would require checking this
+            // flags in a single setcc operation, namley LessThan and LessThanOrEqual.
+            // Because setcc does not check together the PF with either ZF or CF
+            // we swap operands for condition codes that would benefit by , namely
+            // UnorderedGreaterThan and UnorderedGreaterThanOrEqual
+            match condcode {
+                FloatCC::LessThan
+                | FloatCC::LessThanOrEqual
+                | FloatCC::UnorderedOrGreaterThan
+                | FloatCC::UnorderedOrGreaterThanOrEqual => {
+                    //ctx.emit(Inst::xmm_cmp_rm_r(SseOpcode::Ucomiss, rhs, lhs));
+                    //condcode = condcode.reverse();
+                    unimplemented!("Reverse operands to account for condition code\n");
+                }
+                _ => {
+                    ctx.emit(Inst::xmm_cmp_rm_r(SseOpcode::Ucomiss, lhs, rhs));
+                }
+            }
+
+            // Setcc checks to or fewer of these flags at a time, therefore there
+            // we cannot use a single setcc to determine any of these condition codes.
+            // Instead we load the flags into
```

Comment:
nit: unfinished comment here.

It seems the old backend would generate two comparisons + setcc for equality and non-equality, and `and` the results; are these the two only cases where we can't just use a single `setcc` instructions?
Alternatively, it seems that Spidermonkey will do an inline test for the parity flag and jump inline to set the destination result to the final value in this case. Doing likewise would require this instruction to become a synthetic sequence, so the branch and jump aren't separated by a spill/reload.

Reply:
> nit: unfinished comment here.
> 
> It seems the old backend would generate two comparisons + setcc for equality and non-equality, and `and` the results; are these the two only cases where we can't just use a single `setcc` instructions?


I saw that about non-equality and was confused by it. Non-equality should require a check that the ZF is set to 0. If that is the case then the parity bit cannot be set. From what I see the only case supported in cranelift that needs a separate check is the equality condition. The llvm code that I was looking at does have a case for unordered or non-equality that I think also requires a separate check. 

> Alternatively, it seems that Spidermonkey will do an inline test for the parity flag and jump inline to set the destination result to the final value in this case. Doing likewise would require this instruction to become a synthetic sequence, so the branch and jump aren't separated by a spill/reload.

Not quote sure I follow here but would like to see this code to understand. I had considered that a cmov would potentially be useful here but it doesn't seem jmps or cmovs will be necessary.